Python: Releases newer than 0.8.6 will not install on docker alpine

<!-- Hey there, thank you for creating an issue! -->

**Describe the bug you encountered:**

I am running the latest python:3.12-alpine docker image and I am unable to install the pip package newer than 0.8.6.
...

**What did you expect to happen instead?**

If I run `pip install pyroscope-io==0.8.8` I should be able to install the python whl file. This is on `x86_64`

```
/ # pip install pyroscope-io==0.8.8
ERROR: Could not find a version that satisfies the requirement pyroscope-io==0.8.8 (from versions: 0.3.0, 0.3.1, 0.3.2, 0.4.0, 0.4.1, 0.4.2, 0.4.3, 0.4.4, 0.4.5, 0.4.6, 0.4.7, 0.5.0, 0.5.1, 0.6.0, 0.7.0, 0.7.1, 0.7.2, 0.8.0, 0.8.1, 0.8.2, 0.8.3, 0.8.4, 0.8.5, 0.8.6)
ERROR: No matching distribution found for pyroscope-io==0.8.8
/ # pip install pyroscope-io==0.8.10
ERROR: Could not find a version that satisfies the requirement pyroscope-io==0.8.10 (from versions: 0.3.0, 0.3.1, 0.3.2, 0.4.0, 0.4.1, 0.4.2, 0.4.3, 0.4.4, 0.4.5, 0.4.6, 0.4.7, 0.5.0, 0.5.1, 0.6.0, 0.7.0, 0.7.1, 0.7.2, 0.8.0, 0.8.1, 0.8.2, 0.8.3, 0.8.4, 0.8.5, 0.8.6)
ERROR: No matching distribution found for pyroscope-io==0.8.10
```
...


**How did you install `pyroscope-rs`?**

I used the command above and pip install command

---

**pyroscope-rs version and environment**

> 0.8.6, docker.io/python:3.12-alpine

@jmcf-pf Your initial description is about not being able to install packages newer than 0.8.6 in Alpine, but then your third comment shows that it also fails to install 0.8.6. Did you ever manage to install any version of pyroscope-io on Alpine?

I end up with the same issue about `ffikit` when trying to install 0.8.6, and with the issue about missing sources using a higher version.

<details><summary>I've tried to build it directly from the repository as suggested by @bmccool in the other issue, but am then faced with this error:</summary>

```
 > [pyroscope-build 4/4] RUN --mount=type=cache,target=/root/.cargo/registry     --mount=type=cache,target=/root/.cargo/git     --mount=type=cache,target=/root/.cache/uv     cd pyroscope-rs/pyroscope_ffi/python &&     uv run --no-project --with build make build &&     mv dist /dist:  
0.309 rm -rf build                                                                                                                                                                                                                                                                           
0.310 rm -rf dist                                                                                                                                                                                                                                                                            
0.310 rm -rf wheelhouse                                                                                                                                                                                                                                                                      
0.311 rm -rf pyroscope_beta.egg-info                                                                                                                                                                                                                                                         
0.311 rm -rf pyroscope_io.egg-info
0.311 rm -rf .eggs
0.312 rm -rf lib/target
0.312 python -m build --wheel
0.830 * Creating isolated environment: venv+pip...
0.846 * Installing packages in isolated environment:
3.049   - pyromilksnakex==0.1.8
3.049   - setuptools>=42
3.049   - wheel
3.049 * Getting build dependencies for wheel...
3.342 /tmp/build-env-sgdt634w/lib/python3.13/site-packages/setuptools/dist.py:761: SetuptoolsDeprecationWarning: License classifiers are deprecated.
3.342 !!
3.342 
3.342         ********************************************************************************
3.342         Please consider removing the following classifiers in favor of a SPDX license expression:
3.342 
3.342         License :: OSI Approved :: Apache Software License
3.342 
3.342         See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
3.342         ********************************************************************************
3.342 
3.342 !!
3.342   self._finalize_license_expression()
3.366 * Installing packages in isolated environment:
3.980   - pyromilksnakex==0.1.8
3.980 * Building wheel...
4.272 /tmp/build-env-sgdt634w/lib/python3.13/site-packages/setuptools/dist.py:761: SetuptoolsDeprecationWarning: License classifiers are deprecated.
4.272 !!
4.272 
4.272         ********************************************************************************
4.272         Please consider removing the following classifiers in favor of a SPDX license expression:
4.272 
4.272         License :: OSI Approved :: Apache Software License
4.272 
4.272         See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
4.272         ********************************************************************************
4.272 
4.272 !!
4.272   self._finalize_license_expression()
4.272 running bdist_wheel
4.282 running build
4.282 running build_py
4.282 creating build/lib.linux-x86_64-cpython-313/pyroscope
4.283 copying pyroscope/__init__.py -> build/lib.linux-x86_64-cpython-313/pyroscope
4.283 creating build/lib.linux-x86_64-cpython-313/pyroscope_io
4.284 copying pyroscope_io/__init__.py -> build/lib.linux-x86_64-cpython-313/pyroscope_io
4.588 error: cannot produce cdylib for `pyroscope_ffi v0.1.0 (/pyroscope-rs/pyroscope_ffi/python/lib)` as the target `x86_64-unknown-linux-musl` does not support these crate types
4.638 
4.638 ERROR Backend subprocess exited when trying to invoke build_wheel
4.656 make: *** [Makefile:13: build] Error 1
```

</details>

<details><summary>Dockerfile:</summary>

```dockerfile
FROM python:3.13-alpine AS build_base
COPY --from=ghcr.io/astral-sh/uv:0.7.2 /uv /bin/uv

RUN apk add --no-cache git curl gcc musl-dev make
RUN (curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y)
ENV PATH="/root/.cargo/bin:${PATH}"

# Build pyroscope-io from source, because of https://github.com/grafana/pyroscope/issues/1796#issuecomment-1777810359
RUN git clone https://github.com/grafana/pyroscope-rs --branch python-0.8.6
RUN --mount=type=cache,target=/root/.cargo/registry \
    --mount=type=cache,target=/root/.cargo/git \
    --mount=type=cache,target=/root/.cache/uv \
    cd pyroscope-rs/pyroscope_ffi/python && \
    uv run --no-project --with build make build && \
    mv dist /dist
```


</details>

--- a/docker/wheel-musllinux.Dockerfile
+++ b/docker/wheel-musllinux.Dockerfile
@@ -1,0 +1,42 @@
+# syntax=docker/dockerfile:1.4
+ARG PLATFORM=x86_64
+FROM quay.io/pypa/musllinux_1_2_${PLATFORM} AS builder
+
+RUN apk add --no-cache gcc musl-dev libffi-dev make
+
+RUN adduser -D builder \
+    && mkdir -p /pyroscope-rs \
+    && chown builder:builder /pyroscope-rs
+
+USER builder
+RUN test "$(id -u)" = "1000" || { echo "ERROR: builder uid is $(id -u), expected 1000"; exit 1; }
+
+ENV RUST_VERSION=1.87
+RUN curl https://static.rust-lang.org/rustup/dist/$(arch)-unknown-linux-musl/rustup-init -o /tmp/rustup-init \
+    && chmod +x /tmp/rustup-init \
+    && /tmp/rustup-init -y --default-toolchain=${RUST_VERSION} --default-host=$(arch)-unknown-linux-musl \
+    && rm /tmp/rustup-init
+ENV PATH=/home/builder/.cargo/bin:$PATH
+
+WORKDIR /pyroscope-rs
+
+RUN /opt/python/cp39-cp39/bin/python -m pip install --user build
+
+ADD --chown=builder:builder pyproject.toml \
+    rustfmt.toml \
+    Cargo.toml \
+    Cargo.lock \
+    ./
+
+ADD --chown=builder:builder src src
+ADD --chown=builder:builder pyroscope_ffi/ pyroscope_ffi/
+
+RUN --mount=type=cache,target=/home/builder/.cargo/registry,uid=1000,gid=1000 \
+    --mount=type=cache,target=/home/builder/.cargo/git,uid=1000,gid=1000 \
+    /opt/python/cp39-cp39/bin/python -m build --wheel
+
+USER root
+RUN auditwheel repair dist/*.whl --wheel-dir dist-repaired/
+
+FROM scratch
+COPY --from=builder  /pyroscope-rs/dist-repaired dist/

--- a/ffi.mk
+++ b/ffi.mk
@@ -22,6 +22,24 @@ wheel/linux/arm64:
 	 	-f docker/wheel.Dockerfile \
 	 	.
 
+.phony: wheel/musllinux/amd64
+wheel/musllinux/amd64:
+	docker buildx build \
+		--build-arg=PLATFORM=x86_64 \
+	 	--platform=linux/amd64 \
+	 	--output=. \
+	 	-f docker/wheel-musllinux.Dockerfile \
+	 	.
+
+.phony: wheel/musllinux/arm64
+wheel/musllinux/arm64:
+	docker buildx build \
+		--build-arg=PLATFORM=aarch64 \
+	 	--platform=linux/arm64 \
+	 	--output=. \
+	 	-f docker/wheel-musllinux.Dockerfile \
+	 	.
+
 .phony: wheel/mac/amd64
 wheel/mac/amd64:
 	MACOSX_DEPLOYMENT_TARGET=11.0 CARGO_BUILD_TARGET=x86_64-apple-darwin python -m build --wheel

--- a/src/utils.rs
+++ b/src/utils.rs
@@ -29,6 +29,11 @@ pub struct ThreadId {
     pthread: libc::pthread_t,
 }
 
+// SAFETY: pthread_t is an opaque thread identifier used as a handle,
+// never dereferenced. On musl it's *mut c_void, on glibc it's c_ulong.
+unsafe impl Send for ThreadId {}
+unsafe impl Sync for ThreadId {}
+
 impl From<libc::pthread_t> for ThreadId {
     fn from(value: libc::pthread_t) -> Self {
         Self { pthread: value }
@@ -42,7 +47,7 @@ impl ThreadId {
     }
 
     pub fn to_string(&self) -> String {
-        self.pthread.to_string()
+        (self.pthread as usize).to_string()
     }
 }
 

